TMU_Testing/TMU.c

```diff
@@ -0,0 +1,275 @@
+               case TIMER_2:
+                      instance.Timer_CH=TIMER_2;
+                      instance.Timer_Mode=TIMER_MODE;
+                      instance.Timer_Int_MODE=TIMER_INTERRUPT_MODE;
+                      instance.Timer_Prescaler=TIMER_PRESCALER_64;
+                      Timer_Init(&instance);
+                      break;
+               default:
+                      status =E_NOK;
+                      break;
+            }
+          }
+      }
+   return status;
+}
+
+
+/*
+Function Name 	TMU_Start_Timer
+Syntax:
+Sync/Async:
+Reentrancy:
+Parameters (in):
+Parameters (out):
+Parameters (inOut):
+Return:*/
+
+EnmTMUError_t TMU_Start_Timer(uint16_t time,ptrtotask_CBK_FUNC ptrtofun,uint8_t periodicORoneshot)
+{   EnmTMUError_t status=E_OK;
+   if(gu8_multiple_init==not_init)
+   {
+      status=TMU_MODULE_NOT_INIT;
+   }
+else
+{
+         if(gs_arraycount<BUFFER_SIZE)
+         {
+            if((periodicORoneshot!=PERIODIC)&&(periodicORoneshot!=ONESHOT))
+            {
+               status=TMU_NOT_PERODICORONESHOT;
+            }
+            else
+            {
+              if(ptrtofun==NULL)
+              {
+                 status=TMU_FUN_NULL_POINTER;
+              }
+              else
+              {
+                    gstr_arrayoftasks[gs_arraycount].delay=time;
+                    gstr_arrayoftasks[gs_arraycount].Ptrtotask=ptrtofun;
+                    gstr_arrayoftasks[gs_arraycount].perodicity=periodicORoneshot;
+                    gstr_arrayoftasks[gs_arraycount].delay_milistone=CLEAR;
+  
+                    if(gs_arraycount==CLEAR)
+                    {
+                       Timer_Start(TMU_TIMER_ID,TCNT_INITIAL_VALUE);
+                    }
+
+                  gs_arraycount++;
+              }
+            }
+         }
+         else
+         {
+            status=TMU_BUFFER_IS_FULL;
+         }
+}
+return status;
+}
+/**
+* @brief: TMU Stop
+* @param: ptrFun: pointer to callback function
+* @return: EnmTMUError_t status code with error code if one occurred
+*/
+EnmTMUError_t TMU_Stop_Timer(void (*ptrtofun)(void))
+{
+   EnmTMUError_t status=E_OK;
+   uint8_t au8_counter=CLEAR;
+   uint8_t Flag=CLEAR;
+   if(gs_arraycount==CLEAR)
+   {
+      status=TMU_BUFFER_IS_EMPTY;
+   }
+   else
+  {
+         for(au8_counter=START;au8_counter<BUFFER_SIZE;au8_counter++)
+         {
+
+            if(gstr_arrayoftasks[au8_counter].Ptrtotask==ptrtofun)
+            {   Flag=HIGH;
+               if(au8_counter==BUFFER_SIZE-ONE)
+               {
+                  gstr_arrayoftasks[au8_counter].Ptrtotask=NULL;
+                  gstr_arrayoftasks[au8_counter].delay_milistone=CLEAR;
+                  gstr_arrayoftasks[au8_counter].delay=CLEAR;
+                  gstr_arrayoftasks[au8_counter].perodicity=CLEAR;
+               }
+               else
+               {
+                  gstr_arrayoftasks[au8_counter].Ptrtotask=gstr_arrayoftasks[BUFFER_SIZE-ONE].Ptrtotask;
```

Comment:
I didn’t understand why we have two arrays !!!

Reply:
hi hossam thanks for your comment i agreed for all other comments but here i don't have two array it just one array in this line i swap the last element of the array of structure with the element i want to delete it from the array after that  decrements the array counter and null init the last element if you still doesn't understand we can talk in it next session